### PR TITLE
Remove python 3.5 support. Fixes #112

### LINKED ISSUE #112: Drop support for Python 3.5
`wsproto` merged a change last month that drops support for Python 3.5: https://github.com/python-hyper/wsproto/pull/116

This change hasn't been released yet, but it is planned to be part of their version 1.0  (tentatively scheduled for October 2019), so in order for us to consume `wsproto==1.0`, this project will need to drop Python 3.5 sometime around the end of 2019.

When we do this, we can close #54 and #84, and also remove the `async_generator` dependency.

I'm opening this thread now even though I'm not going to make any changes until `wsproto` forces us to. Please leave comments if you have concerns.

### PR DESCRIPTION
**Pull request by AutomatedTester**

### CODE REVIEW

**Comment by AutomatedTester** [comment]
Failures due to #137

**Comment by belm0** [comment]
this would also need:
  * remove async_generator dependency as mentioned in #112 
  * update minimum Python version in README

File: setup.py [hunk 1]
```diff
@@ -28,17 +28,15 @@
...
         'Programming Language :: Python :: 3.7',
         'Programming Language :: Python :: Implementation :: CPython',
         'Programming Language :: Python :: Implementation :: PyPy',
     ],
-    python_requires=">=3.5",
+    python_requires=">=3.6",
     keywords='websocket client server trio',
     packages=find_packages(exclude=['docs', 'examples', 'tests']),
     install_requires=[
-        'async_generator>=1.10,<2',
```
Comment (belm0) [inline setup.py:41]: please remove the code references to the async_generator module

**Comment by AutomatedTester** [comment]
Done

* [x]   remove async_generator dependency as mentioned in #112
* [x]  update minimum Python version in README
* [x]  update python_requires in setup.py

**Comment by AutomatedTester** [comment]
Hey @belm0 is there anything left to do to get this merged?

**Comment by belm0** [comment]
> remove async_generator dependency as mentioned in #112

It means removing all references to this 3rd party module:

https://github.com/HyperionGray/trio-websocket/blob/d878659a753bb4ecc97f5b39485b9af77fdf639a/trio_websocket/_impl.py#L9

https://github.com/HyperionGray/trio-websocket/blob/be300518ff33b18d8c2d7a8b85ad3bd15023457a/tests/test_connection.py#L40

https://github.com/HyperionGray/trio-websocket/blob/b798551ad562ef24000859c6710665818c77135a/setup.py#L42

In Python 3.6, the `yield` keyword can be used normally within an async generator, and so `async_generator` and `yield_` are not needed.  And use `asynccontextmanager` from the stdlib.

**Comment by ehaas** [comment]
> In Python 3.6, the `yield` keyword can be used normally within an async generator, and so `async_generator` and `yield_` are not needed. And use `asynccontextmanager` from the stdlib.

FYI `asynccontextmanager` requires 3.7 -- see https://docs.python.org/3/library/contextlib.html#contextlib.asynccontextmanager  "New in version 3.7."

**Comment by belm0** [comment]
> FYI `asynccontextmanager` requires 3.7 -- see https://docs.python.org/3/library/contextlib.html#contextlib.asynccontextmanager "New in version 3.7."

Sorry, my oversight!

For this PR, removing use of the `@async_generator` decorator and replacing `yield_` with `yield` look straightforward.  I'd like to do that, and restore the dependency on `async_generator` which is needed for the asynccontextmanager shim.

(Later we can consider implementing the async context manager manually to remove the dependency on this lib.)

**Comment by ehaas** [comment]
In case it's helpful - I extracted `asynccontextmanager` from latest cpython and removed some unnecessary abstraction. Tested in Python 3.6.11 and it seems to work. Only issue is that by default my linter complained about `Async context manager 'async_generator' doesn't implement __aenter__ and __aexit__ (not-async-context-manager)` when using it to wrap a function. The code itself works though.

```python3
from functools import wraps

class _AsyncGeneratorContextManager:
    """Helper for @asynccontextmanager."""

    def __init__(self, func, args, kwds):
        self.gen = func(*args, **kwds)
        self.func, self.args, self.kwds = func, args, kwds
        # Issue 19330: ensure context manager instances have good docstrings
        doc = getattr(func, "__doc__", None)
        if doc is None:
            doc = type(self).__doc__
        self.__doc__ = doc
        # Unfortunately, this still doesn't provide good help output when
        # inspecting the created context manager instances, since pydoc
        # currently bypasses the instance docstring and shows the docstring
        # for the class instead.
        # See http://bugs.python.org/issue19404 for more details.

    async def __aenter__(self):
        try:
            return await self.gen.__anext__()
        except StopAsyncIteration:
            raise RuntimeError("generator didn't yield") from None

    async def __aexit__(self, typ, value, traceback):
        if typ is None:
            try:
                await self.gen.__anext__()
            except StopAsyncIteration:
                return
            else:
                raise RuntimeError("generator didn't stop")
        else:
            if value is None:
                value = typ()
            # See _GeneratorContextManager.__exit__ for comments on subtleties
            # in this implementation
            try:
                await self.gen.athrow(typ, value, traceback)
                raise RuntimeError("generator didn't stop after athrow()")
            except StopAsyncIteration as exc:
                return exc is not value
            except RuntimeError as exc:
                if exc is value:
                    return False
                # Avoid suppressing if a StopIteration exception
                # was passed to throw() and later wrapped into a RuntimeError
                # (see PEP 479 for sync generators; async generators also
                # have this behavior). But do this only if the exception wrapped
                # by the RuntimeError is actully Stop(Async)Iteration (see
                # issue29692).
                if isinstance(value, (StopIteration, StopAsyncIteration)):
                    if exc.__cause__ is value:
                        return False
                raise
            except BaseException as exc:
                if exc is not value:
                    raise

def asynccontextmanager(func):
    @wraps(func)
    def helper(*args, **kwds):
        return _AsyncGeneratorContextManager(func, args, kwds)
    return helper

```

**Comment by belm0** [comment]
@AutomatedTester would you open this PR for maintainer edits?

I have some follow-up commits pending for changes in master and using native async generators.

Screen Shot 2020-11-21 at 3 49 07 PM

**Comment by AutomatedTester** [comment]
> @AutomatedTester would you open this PR for maintainer edits?
> 
> I have some follow-up commits pending for changes in master and using native async generators.
> 
> Screen Shot 2020-11-21 at 3 49 07 PM

It showing as already done

**Pull request merged**